tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py

```diff
@@ -375,259 +416,1614 @@ def test_bal_extcodesize_and_oog(
 
 
 @pytest.mark.parametrize(
-    "fails_at_call", [True, False], ids=["oog_at_call", "successful_call"]
+    "target_is_warm", [False, True], ids=["cold_target", "warm_target"]
+)
+@pytest.mark.parametrize(
+    "target_is_empty", [False, True], ids=["existing_target", "empty_target"]
+)
+@pytest.mark.parametrize("value", [0, 1], ids=["no_value", "with_value"])
+@pytest.mark.parametrize(
+    "memory_expansion", [False, True], ids=["no_memory", "with_memory"]
 )
-def test_bal_call_and_oog(
+def test_bal_call_success_no_delegation(
     pre: Alloc,
     blockchain_test: BlockchainTestFiller,
     fork: Fork,
-    fails_at_call: bool,
+    target_is_warm: bool,
+    target_is_empty: bool,
+    value: int,
+    memory_expansion: bool,
+    call_warmup_bytecode_cost: int,
 ) -> None:
-    """Ensure BAL handles CALL and OOG during CALL appropriately."""
-    alice = pre.fund_eoa()
-    bob = pre.fund_eoa()
+    """CALL without 7702 delegation - test successful execution."""
     gas_costs = fork.gas_costs()
+    alice = pre.fund_eoa()
 
-    # Create contract that attempts to call Bob
-    call_contract_code = Bytecode(
-        Op.PUSH1(0)  # retSize
-        + Op.PUSH1(0)  # retOffset
-        + Op.PUSH1(0)  # argsSize
-        + Op.PUSH1(0)  # argsOffset
-        + Op.PUSH1(0)  # value
-        + Op.PUSH20(bob)  # address
-        + Op.PUSH2(0xFFFF)  # gas (provide enough for the call)
-        + Op.CALL  # Call (cold account access)
-        + Op.STOP
+    target = (
+        pre.empty_account()
+        if target_is_empty
+        else pre.deploy_contract(code=Op.STOP)
     )
 
-    call_contract = pre.deploy_contract(code=call_contract_code)
+    # warmup code if needed
+    warmup_code = Bytecode()
+    if target_is_warm:
+        warmup_code = Op.POP(Op.EXTCODESIZE(target))
```

Comment:
BAL will always have `target` if warmed up via `EXTCODESIZE`. Which means this check will true even if the `CALL` fails:

```sh
 account_expectations[target] = BalAccountExpectation.empty()
```

This is evident for warm up + no value transfer, we cannot determine if the `CALL` went through.

You can set `gas_limit` to `gas_limit -1` and this combination will still return a false positive (even though `CALL` goes OOG).

```sh
uv run fill 'tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py::test_bal_call_success_no_delegation' --watch --clean --until Amsterdam --traces --evm-dump-dir logs -k "no_value and warm_target"
```

Reply:
Yes and this could also open up more edge cases. We need tests which put the targets in the 2930 access list and we need cases where it needs to be added to BAL and not.
To "warm up" an account we should thus do this via the access list and not inside EVM, because it's a good point, this will put them in the BAL.

Reply:
Agreed, used access lists here. Very nice idea! ❤️